Enable WebGL and Canvas2D on Castanets.

Sunwoo-Nam opened the pull request:
To enable WebGL, shared memories of the result, bucket and some data
need to be synchronized between GPU process and remote renderer process.
FYI, I tried to run some webgl sites (Aquarium, Toonshading) for test and
the performance was up to 30fps.

Sunwoo-Nam committed:
Enable WebGL on Castanets.

To enable WebGL, shared memories of the result, bucket and some data
need to be synchronized between GPU process and remote renderer process.
FYI, I tried to run some webgl sites (Aquarium, Toonshading) for test and
the performance was up to 30fps.

Signed-off-by: Sunwoo Nam <sunny.nam@samsung.com>

yg48 commented on gpu/command_buffer/client/implementation_base.cc:
```diff
@@ -207,6 +232,14 @@ bool ImplementationBase::GetBucketContents(uint32_t bucket_id,
+#if defined(CASTANETS)
+  // Request to synchronize shared memory of transfer buffer to get bucket data.
+  if (std::string("renderer") ==
+      base::CommandLine::ForCurrentProcess()->GetSwitchValueASCII("type")) {
```
How about adding function for this? 
i.e.> IsRendererProcess(), IsBrowserProcess() ...

Sunwoo-Nam replied:
It's a great idea. I'll try to make another patch for it because this function has been applied to other patches before.

yg48 replied:
OK :)

suyambulingamrm replied:
Just a clarification, RequestSyncTransferBuffer() is needed both here and WaitForCmd() as well ?

hyowon replied:
> i.e.> IsRendererProcess(), IsBrowserProcess() ...

It can be a good way for code cleanup. But I think it is necessary to make it work properly without checking the process type anywhere.

Sunwoo-Nam replied:
@suyambulingamrm Sorry, I missed your question. Sync of WaitForCmd is for the result of commands and here is for bucket data. So both of them are needed.
But this request sync message was removed and changed on latest PR.
https://github.com/Samsung/Castanets/pull/162

suyambulingamrm replied:
Thanks for the explanation @Sunwoo-Nam.. I got to know later :)

Sunwoo-Nam committed:
Enable Canvas2D for Castanets

To draw Canvas2D contents, GLBuffer is used on skia gl backend. GLBuffer has
a shared memory as memory chunk and it needs to be synchronized after written.

Signed-off-by: Sunwoo Nam <sunny.nam@samsung.com>

Sunwoo-Nam commented:
Enable Canvas2D patch was committed additionally.

hyowon commented on gpu/command_buffer/client/mapped_memory.h:
```diff
@@ -88,6 +93,12 @@ class GPU_EXPORT MemoryChunk {
+#if defined(CASTANETS)
+    if (std::string("renderer") ==
+      base::CommandLine::ForCurrentProcess()->GetSwitchValueASCII("type")) {
+      mojo::SyncSharedMemoryHandle(shm_->backing()->GetGUID(), 0, GetLargestFreeSizeWithWaiting());
```
Check offset size.

Sunwoo-Nam replied:
Done.

hyowon replied:
Does this always sync data from the beginning of the buffer?

Sunwoo-Nam replied:
Sorry, I modificated this line checking the offset, but it seems it was not applied. I'll retry it.

hyowon reviewed (approved):
Adding a new IPC message like GpuChannelMsg_RequestSyncTransferBuffer to control sync of data is not a proper approach. The introduction of a unified sync interface is to avoid adding this kind of code. It also causes repeated synchronization delay, which is not good for performance.
​
However, 2D/3D canvas works well now with this PR. I believe that subsequent refactoring & performance tasks will resolve these problems.​
